domain/src/test/java/org/oppia/android/domain/hintsandsolution/HintsAndSolutionDebugModuleTest.kt

```diff
@@ -0,0 +1,91 @@
+package org.oppia.android.domain.hintsandsolution
+
+import android.app.Application
+import android.content.Context
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.common.truth.Truth.assertThat
+import dagger.Binds
+import dagger.BindsInstance
+import dagger.Component
+import dagger.Module
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.oppia.android.testing.TestLogReportingModule
+import org.oppia.android.testing.robolectric.RobolectricModule
+import org.oppia.android.testing.threading.TestDispatcherModule
+import org.oppia.android.util.data.DataProvidersInjector
+import org.oppia.android.util.data.DataProvidersInjectorProvider
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/** Tests for [HintsAndSolutionDebugModule]. */
+@Suppress("FunctionName")
+@RunWith(AndroidJUnit4::class)
+@LooperMode(LooperMode.Mode.PAUSED)
+@Config(application = HintsAndSolutionDebugModuleTest.TestApplication::class)
+class HintsAndSolutionDebugModuleTest {
```

Comment:
One way you could significantly simplify your debug impl testing surface (if we go with the factory approach) is to actually do the factory selection in the module and test that here. Then, that becomes as simple as:
- Test debug impl provided if setting is off
- Test prod impl is provided if setting is on

Then the debug testing surface is fairly trivial since it's just making sure each of the methods behave correctly in the debug setting. We can assume that the prod impl is correct when the setting is off since this module guarantees that scenario, so no new testing is needed.

Reply:
I have gone with the factory approach but I wasn't able to simplify the tests as I had to create the handler in each test rather than creating it once in the `setUp` function. This is because now the handler implementation is selected based on the setting so first, we have to specify whether the setting is enabled or disabled and then create the hint handler. Is this correct or am I missing something here?

Reply:
I think you can simplify it by:
- Updating the debug module to do the if check & return either the debug or prod version
- Updating these tests correspondingly to verify the correct impl is returned based on the controller state
- Updating the hint handler debug impl tests to directly inject the impl factory (similar to what the prod version does) so that it doesn't even go through the module; This lets us focus those tests just on the debug impl behaviors.

Does that help clarify?

Reply:
If we do the if check in module then wouldn't it provide the implementation on compile time which will always be prod impl. I think we need to do the if check in the debug factory to provide correct implementation at runtime? WDYT? Correct me if I'm wrong.

Reply:
However, I'll try this module approach and let you know if we get the desired result

Reply:
Unfortunately, this didn't work. It is only providing prod impl as I said before. Should we go with the factory approach instead?
![Screenshot from 2021-08-21 15-15-56](https://user-images.githubusercontent.com/63469151/130317990-ac5eb49f-5e8e-494c-9375-466e3c8cca80.png)

Reply:
Dagger arranges the graph at compile time but the providers are actually called at runtime. Since the Factory isn't scoped as singleton, each time it's injected Dagger will re-call the provider.

However, you do bring up a point: we inject the factory just once in the progress controllers which means they won't change without recreating the controllers. The factory approach will work better without broader changes, and your latest test changes seem well in-line with my thoughts on reducing PR complexity.